### PR TITLE
Enhance the reboot method

### PR DESCRIPTION
**Pull request by vvolam**
### Description of PR
Implement a clear method to determine conclusively whether the device has been rebooted.

Summary:
Enhance reboot test

### Type of change
- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311

#### What is the motivation for this PR?
To confirm that the device has been rebooted without encountering any platform or PDU controller issues.

#### How did you do it?
Created a test file in RAM filesystem, which should be deleted after the device is rebooted successfully.

#### How did you verify/test it?
Verified all the tests in tests/platform_tests/test_reboot.py in the lab testbed.

### CODE REVIEW

**Comment by prgeor** [comment]
@vvolam please test negative test case too

**Comment by prgeor** [comment]
@saiarcot895 @vaibhavhd could you also review?

**Comment by vvolam** [comment]
> @vvolam please test negative test case too

Yes @prgeor, I have set the file_check to 0 at runtime using PDB, which made to hit the condition and made sure exception is raised.

**Review by saiarcot895 — APPROVED** [review]

File: tests/common/reboot.py [hunk 1]
```diff
@@ -277,6 +281,12 @@ def reboot(duthost, localhost, reboot_type='cold', delay=10,
         if not ret:
             raise Exception('warmboot-finalizer service timeout on DUT {}'.format(hostname))
 
+    # Verify if the temporary file created in tmpfs is deleted after reboot, to determine a
+    # successful reboot
+    file_check = duthost.command('find /dev/shm/test_reboot', module_ignore_errors=True)['rc']
+    if not file_check:
+        raise Exception('DUT {} reboot did not finish successfully'.format(hostname))
```
Comment (saiarcot895) [inline tests/common/reboot.py:288]: It might be better to reword this as "DUT {} did not reboot" or "DUT {} did not appear to reboot".

Comment (vvolam) [reply tests/common/reboot.py]: Modified it. Thank you!

File: tests/common/reboot.py [hunk 2]
```diff
@@ -241,6 +241,10 @@ def reboot(duthost, localhost, reboot_type='cold', delay=10,
     except KeyError:
         raise ValueError('invalid reboot type: "{} for {}"'.format(reboot_type, hostname))
 
+    # Create a temporary file in tmpfs before reboot
+    logger.info('DUT {} create a file /dev/shm/test_reboot before rebooting'.format(hostname))
+    duthost.command('sudo touch /dev/shm/test_reboot')
```
Comment (vaibhavhd) [inline tests/common/reboot.py:246]: Minor: it may be a good idea to verify that this `touch` command was successful. If the FS is readonly, the issue can be caught right away?

File: tests/common/reboot.py [hunk 3]
```diff
@@ -277,6 +281,12 @@ def reboot(duthost, localhost, reboot_type='cold', delay=10,
         if not ret:
             raise Exception('warmboot-finalizer service timeout on DUT {}'.format(hostname))
 
+    # Verify if the temporary file created in tmpfs is deleted after reboot, to determine a
+    # successful reboot
+    file_check = duthost.command('find /dev/shm/test_reboot', module_ignore_errors=True)['rc']
```
Comment (vaibhavhd) [inline tests/common/reboot.py:286]: Why use ` module_ignore_errors=True`? You do care about the errors in this call, don't you?

Comment (vaibhavhd) [inline tests/common/reboot.py:286]: Instead of using your own way of checking for file, you could use preexisting method that comes with Ansible:

```
file_stat = duthost.stat(path="/dev/shm/test_reboot")
pytest_assert(not file_stat["stat"]["exists"], "Test file was found on DUT after reboot")
```

File: tests/common/reboot.py [hunk 4]
```diff
@@ -277,6 +281,12 @@ def reboot(duthost, localhost, reboot_type='cold', delay=10,
...
+    if not file_check:
+        raise Exception('DUT {} did not reboot'.format(hostname))
```
Comment (vaibhavhd) [inline tests/common/reboot.py:288]: It is best to remove the file (cleanup) that you created at this point when the test has failed.

File: tests/common/reboot.py (hunk 3 above)
Comment (vaibhavhd) [reply tests/common/reboot.py:286]: Examples: https://github.com/sonic-net/sonic-mgmt/blob/80415ad3f79c4851091a779bbe2f06c776c929af/tests/common/helpers/backend_acl.py#L26

https://github.com/sonic-net/sonic-mgmt/blob/80415ad3f79c4851091a779bbe2f06c776c929af/tests/bgp/test_bgp_sentinel.py#L305

Comment (vvolam) [reply tests/common/reboot.py:286]: yes, in successful case this command with return error only which we need. If we do not use module_ignore_errors, ansible does not save the error output.

File: tests/common/reboot.py (hunk 2 above)
Comment (vvolam) [reply tests/common/reboot.py:246]: https://www.kernel.org/doc/Documentation/filesystems/tmpfs.txt - As per this kernel documentation, /dev/shm will be always mounted for 2.3 and above kernel versions and this /dev/shm is always mounted as a read-write filesystem in memory.

File: tests/common/reboot.py (hunk 4 above)
Comment (vvolam) [reply tests/common/reboot.py:289]: Incorporated the change. Thank you.

File: tests/common/reboot.py (hunk 3 above)
Comment (vvolam) [reply tests/common/reboot.py]: Modified the PR to use the existing method. Thank you!

**Review by vaibhavhd — APPROVED** [review]

File: tests/common/reboot.py (hunk 4 above)
Comment (prgeor) [reply tests/common/reboot.py:289]: @vvolam checked with Vaibhav, don't delete this file explicitly for debugging purpose

**Review by prgeor — APPROVED** [review]
